### ISSUE TITLE
Core module for communication with AI

### ISSUE THREAD

**Issue by kamran-tuni** [kamran-tuni/smart-rule-engine#1]
Test Cases:

- Send system prompt, system data and expected JSON to AI and verify it's readiness.
- Send user prompt with missing info and verify that AI asks for missing info.
- Provide missing info and verify that AI provides required JSON schema for Rule Chain.

Enum:

NodeTypes

```
SOURCE_NODE = source_node
SCRIPT_NODE = script_node
SWITCH_NODE = switch_node
ACTION_NODE = action_node
```

Entity:

SourceNodeConfig
```
device_id: int
parameter_id: int
```

ScriptNodeConfig
```
script: str
```

SwitchNodeConfig
```
condition: str
value: [int, str]
```

ActionNodeConfig
```
device_id: int 
parameter_id: int
value: [int, str]
```

RuleChain
```
nodes: list[Node]
```

Node
```
id: uuid
name: str
type: NodeTypes
config: [SourceNodeConfig, ScriptNodeConfig, SwitchNodeConfig, ActionNodeConfig]
target_node_id: uuid
```

JSON config file: `config.py`

```
"system_prompt": ...
"system_data": ....
"expected_rule_chain": ... 
```

UseCase:

InitAIUseCase
```
Input: None
Output: None (Exception is raised in case of failure in initialization)
Dependency Injection: AI client object. Apply singleton pattern for AI client object.
Logic: Read config file and send system prompt, data and expected rule chain to AI and wait for system to be read. When system is ready return. If system not ready raise exception SYSTEM_NOT_READY
```
GenerateRuleChainUseCase
```
Input: user_prompt
Output: Tue, result rule chain json ----- False, Query  
Dependency Injection: AI client object. Apply singleton pattern for AI client object.
Logic: Send user prompt to AI. If AI provides JSON schema in response and then return only the rule chain JSON schema with True Flag. If AI asks for missing info, then return this query in response with False flag.
```
The `GenerateRuleChainUseCase` will be called multiple times based on user prompt and until desired rule chain schema is generated.

UseCase TestCases:

InitAIUseCaseTest

```
Invoke usecase and see if no exception is raised.
Invoke usecase with wrong system prompt and see if exception is raised.
```

GenerateRuleChainUseCaseTest

```
Invoke usecase with missing info and verify response as False, Query
Invoke usecase with missing info and verify response as False, Query, then provide missing info and verify response as True, JSON
Invoke usecase with all info and verify response as True, JSON
```
